op-txproxy/auth_handler.go

```diff
@@ -0,0 +1,95 @@
+package op_txproxy
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var (
+	defaultBodyLimit = 5 * 1024 * 1024 // default in op-geth
+
+	DefaultAuthHeaderKey = "x-optimism-signature"
+)
+
+// authHandler detects when authentication information is present on the request.
+// If so, it will validate and set authenicated caller in the request context. It does not
+// reject if authentication information is missing. It is up to the request handler to do so.
+type authHandler struct {
+	headerKey string
+	next      http.Handler
+}
+
+func AuthMiddleware(headerKey string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return &authHandler{headerKey, next}
+	}
+}
+
+type authContextKey struct{}
+
+type AuthContext struct {
+	Caller common.Address
+}
+
+// ServeHTTP serves JSON-RPC requests over HTTP, implements http.Handler
+func (h *authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	authHeader := r.Header.Get(h.headerKey)
+	if authHeader == "" {
+		h.next.ServeHTTP(w, r)
+		return
+	}
+	authElems := strings.Split(authHeader, ":")
+	if len(authElems) != 2 {
+		http.Error(w, "misformatted auth header", http.StatusBadRequest)
+		return
+	}
+
+	if r.Body == nil {
+		// edge case from unit tests
+		r.Body = io.NopCloser(bytes.NewBuffer(nil))
+	}
+
+	// The request body can only be read once. Adjust here so it's inspectable
+	var buf bytes.Buffer
+	body := io.TeeReader(r.Body, &buf)
```

Comment:
You forgot to schedule closing the original Request.Body. This would be very bad in prod.

Reply:
ah! good catch